fairmq/shmem/Manager.h

```diff
@@ -71,7 +70,58 @@ struct ShmPtr
 
     char* UserPtr()
     {
-        return realPtr + sizeof(uint16_t) + *(reinterpret_cast<uint16_t*>(realPtr));
+        if (realPtr) {
+            return realPtr + sizeof(uint16_t) * 2 + alignof(std::atomic<uint16_t>) + sizeof(std::atomic<uint16_t>) + *(reinterpret_cast<uint16_t*>(realPtr));
+        } else {
+            return nullptr;
+        }
+    }
+
+    static uint16_t RefCount(char* rPtr)
+    {
+        auto rcPtr = reinterpret_cast<std::atomic<uint16_t>*>(rPtr
+                                                              + sizeof(uint16_t) * 2
+                                                              + *(reinterpret_cast<uint16_t*>(rPtr + sizeof(uint16_t))));
```

Comment:
Could you please add some inline comments that explain what your are doing in those terms in english words? A general comment describing the byte layout of the `ShmHeader` would be really helpful, too. Thx!

Reply:
I have the layout in a comment a bit further in the code, but sure I'll pull it up and add some prose.